### ISSUE TITLE
Flaky dart2js failures due to misoptimized code

### ISSUE THREAD

**Issue by mraleph** [dart-lang/sdk#26558]
We have certain amount of flakiness on the dart2js bots. Suspecting misoptimization in background thread. This issue is to document an effort in finding the source. 

Issues can be reproduced by running 

```
 python -u tools/test.py --mode=release --arch=ia32 --compiler=dart2js --dart2js-batch --runtime=d8 --progress=buildbot -v --clear_browser_cache --report --time --failure-summary --write-debug-log --write-test-outcome-log --copy-coredumps --minified --shards=5 --shard=4 --checked --append_logs
```

a lot of times. One common failure is `NoSuchMethod` due to trying to call `get:kind` on `null` in the method 

```dart
  ConstructorResult visitSend(Send node) {
    ConstructorResult receiver = visit(node.receiver);
    assert(invariant(node.receiver, receiver != null,
        message: 'No result returned for $node.receiver.'));
    if (receiver.kind != null) {
      assert(invariant(node, receiver.element.isMalformed,
          message: "Unexpected prefix result: $receiver."));
      // We have already found an error.
      return receiver;
    }

    Identifier name = node.selector.asIdentifier();
    if (name == null) {
      reporter.internalError(node.selector, 'unexpected node');
    }

    if (receiver.type != null) { /* (!) */
      if (receiver.type.isInterfaceType) {  // NoSuchMethod inside isInterfaceType
 ```

I managed to get a core dump and extract optimized code for this method from it. As far as I can see compiler removed the `null`-check marked as `(!)` in the code above.

```asm
Dump of assembler code from 0xec214d61 to 0xec2151c9:
   0xec214d61:	mov    ebp,esp
   0xec214d63:	push   0xebfe3959
   0xec214d68:	sub    esp,0xc
   0xec214d6b:	mov    eax,DWORD PTR [ebp+0x8]
   0xec214d6e:	cmp    esp,DWORD PTR [esi+0x4]
   0xec214d71:	jbe    -> stack check
   0xec214d77:	test   al,0x1
   0xec214d79:	je     -> deopt
   0xec214d7f:	movzx  ecx,WORD PTR [eax+0x1]
   0xec214d83:	cmp    ecx,0x7b8
   0xec214d89:	jne    -> deopt

   0xec214d8f:	mov    ecx,DWORD PTR [eax+0xb]   ; node.receiver
   0xec214d92:	cmp    ecx,0xf67c0021            ; node.receiver == null ?
   0xec214d98:	je     0xec214ef6
   0xec214d9e:	push   ecx                       ; node.receiver.accept(this)
   0xec214d9f:	push   DWORD PTR [ebp+0xc]
   0xec214da2:	mov    ebx,DWORD PTR [esp+0x4]
   0xec214da6:	mov    ecx,0xef4073a9
   0xec214dab:	mov    edi,0xf67c1e81
   0xec214db0:	call   DWORD PTR [edi+0x3]
   0xec214db3:	call   ebx
   0xec214db5:	add    esp,0x8
   ;;; eax = node.receiver == null ? null : node.receiver.accept(this)
   0xec214db8:	mov    DWORD PTR [ebp-0x10],eax ;
   0xec214dbb:	test   al,0x1
   0xec214dbd:	je     ->deopt
   0xec214dc3:	movzx  ecx,WORD PTR [eax+0x1]
   0xec214dc7:	cmp    ecx,0x9b5
   0xec214dcd:	jne    ->deopt
   0xec214dd3:	mov    ecx,DWORD PTR [eax+0x7] ; ecx = receiver.kind
   0xec214dd6:	cmp    ecx,0xf67c0021          ; ecx == null
   0xec214ddc:	jne    0xec214f00              ; ->return 

   0xec214de2:	mov    ecx,DWORD PTR [ebp+0x8] ; ecx = node
   0xec214de5:	mov    edx,DWORD PTR [ecx+0xf] ; node.selector
   0xec214de8:	mov    DWORD PTR [ebp-0xc],edx ;
   0xec214deb:	test   dl,0x1                  ;
   0xec214dee:	je     ->deopt                 ;
   0xec214df4:	movzx  ebx,WORD PTR [edx+0x1]  ;
   0xec214df8:	cmp    ebx,0x7d0               ;
   0xec214dfe:	jne    ->deopt                 ;

   0xec214e04:	mov    ecx,DWORD PTR [eax+0xf] ; ecx = receiver.type
   0xec214e07:	mov    DWORD PTR [ebp-0x8],ecx
   0xec214e0a:	push   ecx

   0xec214e0b:	mov    ebx,DWORD PTR [esp]     ; get:kind
   0xec214e0e:	mov    ecx,0xf64dc691               
   0xec214e13:	mov    edi,0xf67c1e81
   0xec214e18:	call   DWORD PTR [edi+0x3]     
   0xec214e1b:	call   ebx
```

We suspect that it's again some unforeseen interaction between guarded fields and background compiler, but so far we were not able to figure out precise reasons.

**Comment by a-siva** [comment]
@fsc8000

**Comment by fsc8000** [comment]
Since we have not found the issue yet, I think we should we switch it off on the stable branch for 1.17.

**Comment by mit-mit** [comment]
When was this optimization introduced?

**Comment by mraleph** [comment]
@mit-mit background compiler was introduced pre 1.6 and it was switched off by default in 1.6 stable release IIRC.

**Comment by mit-mit** [comment]
@mraleph and when was it switched on again?

**Comment by mraleph** [comment]
I think here https://github.com/dart-lang/sdk/commit/80ba69dfdb72541f512ea760274ae335618b6f2e

**Comment by fsc8000** [comment]
We can just prepare a one-time CL to do this. No need to merge.
Am 02.06.2016 12:30 schrieb "Vyacheslav Egorov" <notifications@github.com>:

**Comment by fsc8000** [comment]
https://codereview.chromium.org/2035673002

would be the CL for the stable release to disable it. I guess it will have to wait until 1.18 to be on by default.

**Comment by mit-mit** [comment]
CC @whesse

**Comment by mit-mit** [comment]
Can you file a merge request please? We are not doing any more full pushes to dev.

**Comment by mit-mit** [comment]
This was disable for 1.17 stable. Moving this to 1.18 -- we need to get this stabilised for 1.18 stable.

**Comment by mraleph** [comment]
I am now fairly certain that this issue is caused by a bug in parallel marking and not actually in the background compilation itself. It seems parallel marking is handling weak properties incorrectly - sometimes clearing those that should not be cleared (because key is still reachable). We realy on weak properties for dependency tracking - and incorrect handling in GC causes us to loose dependcy between a field and an optimized code object. When field guarded cid state changes (from non-nullable to nullable) we don't deoptimize code object because the dependency was incorrectly erased by the GC.

We need to rewrite how weak properties are handled in the GC. 

Tentatively assigning to @a-siva for triage, I am not sure who owns GC subsystem now. 

@a-siva feel free to reassign back if you would like me to fix the GC.

**Comment by mraleph** [comment]
Weak property processing is fixed in 97a8c4caffd2e758c60a3e4502f07d6886ee0363.

I think the issue should be resolved now. Lets watch the dart2js bots.

**Comment by fsc8000** [comment]
Nice!
